tensorflow_datasets/core/custom_dataset/image_folder.py

```diff
@@ -0,0 +1,168 @@
+        xxz.png
+      label2/
+        xxx.png
+        xxy.png
+        xxz.png
+    split_name/  # Ex: 'test'
+      ...
+  ```
+
+  To use it:
+
+  ```
+  builder = tfds.ImageLabelFolder(root_dir='path/to/manual_dir/')
+  print(builder.info)  # Splits, num examples,... automatically calculated
+  ds = builder.as_dataset(split='split_name', shuffle_files=True)
+  ```
+
+  """
+
+  VERSION = core.Version('2.0.0')
+
+  def __init__(self, root_dir):
+    self._version = str(self.VERSION)
+    self._root_dir = os.path.expanduser(root_dir)
+
+    # dict[split_name][label_name] = list(img_paths)
+    self._split_label_images, labels = _get_split_label_images(self._root_dir)
+
+    self._ds_info = core.DatasetInfo(
+        builder=self,
+        description='Generic image classification dataset.',
+        features=core.features.FeaturesDict({
+            'image': core.features.Image(),
+            'label': core.features.ClassLabel(names=sorted(list(labels))),
+        }),
+        supervised_keys=('image', 'label'),
+    )
+
+    super(ImageLabelFolder, self).__init__(
+        data_dir=self._root_dir, version=self._version, config=None)
+
+    # Calculate splits for DatasetInfo
+    splits = {
+        split_name: sum(len(l)
+                        for l in self._split_label_images[split_name].values())
+        for split_name in self._split_label_images
+    }
+
+    split_dict = core.SplitDict(self.name)
+    for split_name, size in splits.items():
+      split_dict.add(
+          core.SplitInfo(name=split_name, shard_lengths=[size]))
+
+    self._ds_info.update_splits_if_different(split_dict)
+
+    # Reset `_data_dir` as it should not change to DATA_DIR/Version
+    self._data_dir = self._root_dir
+
+  def _info(self):
+    return self._ds_info
+
+  def _download_and_prepare(self, **kwargs):
+    pass
+
+  def download_and_prepare(self, **kwargs):
+    pass
+
+  def _process_ds(self, image_paths):
+    """Yield image and label"""
+    for path in image_paths:
+      path = path.decode()
+      label = path.split(os.path.sep)[-2]
+      img = tf.io.read_file(path)
+      img = tf.image.decode_jpeg(img, channels=3)
+      img = tf.image.convert_image_dtype(img, tf.uint8)
+      yield {'image': img, 'label': self.info.features['label'].str2int(label)}
+
+  def _as_dataset(self, split, shuffle_files=False, **kwargs):
+    """Generate dataset for given split"""
+    del kwargs
+    if split not in self.info.splits.keys():
+      raise ValueError('Split name {} not present in {}'
+                       .format(split, self._split_label_images.keys()))
+
+    imgs = sorted(
+        img for l in self._split_label_images[split].values() for img in l)
+    ds = tf.data.Dataset.from_generator(
+        self._process_ds,
+        output_shapes=self.info.features.shape,
+        output_types=self.info.features.dtype,
+        args=[imgs])
+
+    if shuffle_files:
+      ds = ds.shuffle(len(imgs))
+    return ds
+
+
+def _get_split_label_images(
```

Comment:
Could you add unittest for this function ? Note I think that you can use `tfds.testing.MockFs()` to avoid having to create real files/directories.

Reply:
I had to fix MockFs listdir function a bit.
Making a new PR for the same.

Reply:
Thank you for fixing the bug.